File: backend/scripts/ai-dubbing/compare_japanese.py
"""Private, offline A/B on one failed cue. No job mutation or media publication."""
import argparse
import math
import os
from pathlib import Path

import runtime as rt


def read_comparison_inputs(failure_root, reference_path, language="ja"):
    failure_root = Path(failure_root)
    generation = rt.load_json(failure_root / "generation-attempt.json")
    manifest = rt.load_json(failure_root / "input-profile.json")
    if (failure_root / "quality-failure.json").is_file():
        failure = rt.load_json(failure_root / "quality-failure.json")
    else:
        failure = rt.load_json(failure_root / "error.json")
        if failure.get("code") != "AI_DUBBING_GENERATION_TIMEOUT":
            raise ValueError("Un diagnostic qualité ou un timeout de génération est requis.")
        matches = [cue for cue in manifest.get("dialogueUnits", []) if
                   cue.get("sourceStart") == generation.get("sourceStart")
                   and cue.get("speaker") == generation.get("speaker")
                   and cue.get("text") == generation.get("text")]
        if len(matches) != 1:
            raise ValueError("La réplique du timeout ne peut pas être identifiée sans ambiguïté.")
        failure = {**failure, "cue": matches[0]}
    cue = failure.get("cue") or {}
    speaker = cue.get("speaker")
    if (generation.get("language") != language or generation.get("operation") != "cue"
            or not isinstance(cue.get("text"), str) or not 0 < len(cue["text"]) <= 500
            or not isinstance(speaker, str) or not rt.re.fullmatch(r"SPEAKER_\d{2}", speaker)):
        raise ValueError(f"Le diagnostic doit décrire une réplique complète dans la langue {language}.")
    for name in ("sourceStart", "start", "end"):
        value = cue.get(name)
        if isinstance(value, bool) or not isinstance(value, (int, float)) or not math.isfinite(value) or value < 0:
            raise ValueError("Horodatages du diagnostic invalides.")
    if cue["end"] <= cue["start"]:
        raise ValueError("La fenêtre initiale est vide.")
    archive_profile = generation.get("pipelineVersion")
    allowed_profiles = {rt.V5_BOUNDED_SAMPLES_PROFILE, rt.V5_JAPANESE_IDENTITY_PROFILE, rt.V5_JAPANESE_BOUNDED_PROFILE}
    if (archive_profile not in allowed_profiles
            or generation.get("speaker") != speaker or generation.get("text") != cue["text"]
            or generation.get("sourceStart") != cue["sourceStart"]
            or failure.get("pipelineVersion") != archive_profile
            or manifest.get("pipelineVersion") != archive_profile
            or generation.get("generationConfigHash") != rt.PROFILE_GENERATION_CONFIG_HASHES[archive_profile]
            or manifest.get("generationConfigHash") != generation.get("generationConfigHash")):
        raise ValueError("Les diagnostics ne correspondent pas à la même réplique/version R5.")
    entry = (manifest.get("references") or {}).get(speaker) or {}
    details = generation.get("details") or {}
    reference_path = Path(reference_path).resolve()
    if (not reference_path.is_file() or rt.digest(reference_path) != entry.get("sha256")
            or entry.get("sha256") != details.get("referenceSha256")
            or entry.get("referenceText") != details.get("referenceText")
            or not entry.get("referenceText")):
        raise ValueError("Le WAV fourni n'est pas la référence exacte du diagnostic (empreinte/texte).")
    seed = generation.get("seed")
    if isinstance(seed, bool) or not isinstance(seed, int) or not 0 <= seed < 2**32:
        raise ValueError("Graine de génération absente ou invalide.")
    reference = {"path": reference_path, "text": entry["referenceText"]}
    return cue, reference, generation, manifest


def compare_variant(root, failure_root, reference_path, output, variant, check_only=False):
    cue, reference, generation, manifest = read_comparison_inputs(failure_root, reference_path)
    profile = {"reference-text": rt.V5_BOUNDED_SAMPLES_PROFILE,
               "speaker-identity": rt.V5_JAPANESE_IDENTITY_PROFILE,
               "bounded-identity": rt.V5_JAPANESE_BOUNDED_PROFILE}[variant]
    report = {
        "schemaVersion": 1, "synthetic": True, "diagnosticOnly": True,
        "variant": variant, "pipelineVersion": profile,
        "generationConfigHash": rt.PROFILE_GENERATION_CONFIG_HASHES[profile],
        "cue": cue, "referenceSha256": rt.digest(reference["path"]),
        "referenceText": reference["text"], "sourceSeed": generation["seed"],
        "seedPolicy": "archived-last-seed-plus-index-paired-v1",
        "warning": "Comparaison ciblée, pas une reproduction garantie des anciens WAV ; écoute requise.",
        "attempts": [],
    }
    if check_only:
        return report
    if any(os.environ.get(name, "").lower() not in rt.OFFLINE_VALUES for name in ("HF_HUB_OFFLINE", "TRANSFORMERS_OFFLINE")):
        raise ValueError("La comparaison doit rester hors ligne.")
    if os.environ.get("SAMI_DUBBING_WATCHDOG_PROTOCOL") != "1":
        raise ValueError("Lancer la comparaison via compareJapanese.mjs (supervision obligatoire).")
    install = rt.install_manifest(root)
    voice = install.get("models", {}).get("voice") or install.get("models", {}).get("chatterbox") or {}
    if install.get("voiceEngine") != "qwen3-tts":
        raise ValueError("Cette comparaison nécessite le clone Qwen3-TTS/CUDA.")
    if (voice.get("repo") != manifest.get("voiceModel")
            or not voice.get("revision") or voice.get("revision") != manifest.get("voiceModelRevision")):
        raise ValueError("Le modèle installé n'est pas celui du diagnostic ; comparaison contrôlée refusée.")
    output = Path(output)
    output.mkdir(mode=0o700, parents=True, exist_ok=False)
    rt.write_json(output / "comparison.json", report)
    model = rt.load_voice_model(install, {cue["speaker"]: reference})
    quality_model = rt.load_quality_model(install)
    import random
    import numpy as np
    import torch
    for attempt in range(3):
        # Same seed and sampling controls in both branches; only the prompt differs.
        seed = (generation["seed"] + attempt) % 2**32
        random.seed(seed)
        np.random.seed(seed)
        torch.manual_seed(seed)
        raw = output / f"AI-diagnostic-attempt-{attempt + 1}.wav"
        try:
            waveform, sample_rate = rt.generate_voice(model, cue["text"], "ja", cue["speaker"], reference,
                attempt=attempt, profile=profile, watch_context={
                    "workspace": output, "kind": "cue", "progress": round(attempt / 3 * 90),
                    "source_start": cue["sourceStart"], "seed": seed,
                    "cue_index": 1, "cue_count": 1, "attempt": attempt + 1,
                })
        except rt.DubbingGenerationLimitError as error:
            rt.write_json(output / f"generation-attempt-{attempt + 1}.json", rt.load_json(output / "generation-attempt.json"))
            report["attempts"].append({"attempt": attempt + 1, "seed": seed,
                "state": "rejected-before-decoding", "code": error.code, "message": str(error), "audio": None})
            rt.write_json(output / "comparison.json", report)
            rt.log(f"Comparaison {variant} : tentative {attempt + 1}/3 REFUSÉE avant décodage (fin absente).")
            continue
        rt.save_generated_audio(waveform, sample_rate, raw)
        # Retain each pre-call diagnostic rather than only the last attempt.
        rt.write_json(output / f"generation-attempt-{attempt + 1}.json", rt.load_json(output / "generation-attempt.json"))
        with rt.observe_voice_activity("quality", progress=90, speaker=cue["speaker"],
                source_start=cue["sourceStart"], cue_index=1, cue_count=1, attempt=attempt + 1):
            recognized = rt.transcribe_quality(quality_model, raw, "ja")
        duration = rt.media_duration(raw)
        decision = rt.japanese_quality_decision(cue["text"], recognized, duration, rt.generated_audio_activity(raw))
        report["attempts"].append({
            "attempt": attempt + 1, "seed": seed, "audio": raw.name,
            "generatedDuration": duration, "recognized": recognized,
            "qualityR5R2": decision, "rawCer": rt.character_error_rate(cue["text"], recognized),
        })
        rt.write_json(output / "comparison.json", report)
        rt.log(f"Comparaison {variant} : tentative {attempt + 1}/3, {duration:.2f}s, CER {decision['cer']:.3f}.")
    return report


def main():
    parser = argparse.ArgumentParser(description=__doc__)
    for name in ("root", "failure", "reference", "output"):
        parser.add_argument(f"--{name}", required=True)
    parser.add_argument("--variant", choices=["reference-text", "speaker-identity", "bounded-identity"], required=True)
    parser.add_argument("--check-only", action="store_true")
    args = parser.parse_args()
    result = compare_variant(args.root, args.failure, args.reference, args.output, args.variant, args.check_only)
    # Stdout is a machine-readable report, including during the model-free
    # preflight. JSON escapes remain lossless even in a legacy Windows console.
    # Diagnostic files continue to use explicit UTF-8 via write_json.
    print(rt.json.dumps(result, ensure_ascii=True))


if __name__ == "__main__":
    main()

File: backend/scripts/ai-dubbing/compareJapanese.mjs
import fs from "node:fs";
import path from "node:path";
import { fileURLToPath } from "node:url";
import { parseArgs } from "node:util";
import { runAiDubbingCommand } from "../../services/aiDubbing/commandRunner.js";

export async function main(args = process.argv.slice(2), { language = "ja" } = {}) {
  if (!["ja", "en"].includes(language)) throw new Error("Langue de comparaison inconnue.");
  const english = language === "en";
  const label = english ? "english" : "japanese";
  const { values } = parseArgs({ args, options: {
    failure: { type: "string" }, reference: { type: "string" }, root: { type: "string" },
    "check-only": { type: "boolean", default: false }, help: { type: "boolean" },
    "prepare-only": { type: "boolean", default: false },
    variant: { type: "string" },
  } });
  if (values.help) {
    if (english) {
      console.log('Usage : node scripts/ai-dubbing/compareEnglish.mjs --failure "dossier failure-UUID" --reference "référence originale SPEAKER_XX.wav" [--root "var/ai-dubbing"] [--check-only | --prepare-only] [--variant reference-text|speaker-identity]\nArrêter le worker avant le test GPU. Par défaut : trois tentatives par mode, six WAV diagnostiques privés. Base R5-R3 inchangée, seul le prompt varie. Aucun job ni piste modifié.');
      return;
    }
    console.log('Usage : node scripts/ai-dubbing/compareJapanese.mjs --failure "dossier failure-UUID" --reference "référence originale SPEAKER_XX.wav" [--root "var/ai-dubbing"] [--check-only | --prepare-only] [--variant bounded-identity|speaker-identity|reference-text|identity-ab]\nArrêter le worker avant le test GPU. Trois tentatives par variante. Par défaut : R5-R1/R5-R2 ; bounded-identity : R5-R3 seule ; identity-ab : R5-R2/R5-R3. Aucun job ni piste modifié. --prepare-only copie un petit dossier portable, sans charger les modèles.');
    return;
  }
  if (!values.failure || !values.reference) throw new Error("--failure et --reference sont requis (voir --help).");
  const choices = { "reference-text": ["reference-text"], "speaker-identity": ["speaker-identity"],
    "bounded-identity": ["bounded-identity"], "identity-ab": ["speaker-identity", "bounded-identity"] };
  if (values.variant && (!choices[values.variant] || (english && !["reference-text", "speaker-identity"].includes(values.variant)))) throw new Error("Variante inconnue pour cette comparaison (voir --help).");
  const variants = choices[values.variant] || ["reference-text", "speaker-identity"];
  if (values["check-only"] && values["prepare-only"]) throw new Error("Choisir --check-only ou --prepare-only.");
  const backend = path.resolve(path.dirname(fileURLToPath(import.meta.url)), "../..");
  const root = path.resolve(values.root || path.join(backend, "var/ai-dubbing"));
  const install = JSON.parse(await fs.promises.readFile(path.join(root, "install.json"), "utf8"));
  if (!path.isAbsolute(install.pythonPath || "") || !fs.existsSync(install.pythonPath)) {
    throw new Error("Python local absent du manifeste d'installation du clone.");
  }
  const failure = path.resolve(values.failure), reference = path.resolve(values.reference);
  const script = path.join(backend, `scripts/ai-dubbing/compare_${label}.py`);
  const controller = new AbortController();
  let interrupted = false;
  const stop = () => { interrupted = true; controller.abort(); };
  process.once("SIGINT", stop);
  process.once("SIGTERM", stop);
  const execute = (variant, output, check = false) => runAiDubbingCommand(install.pythonPath, [
    script, "--root", root, "--failure", failure, "--reference", reference,
    "--output", output, "--variant", variant, ...(check ? ["--check-only"] : []),
  ], { cwd: backend, signal: controller.signal, timeoutMs: check ? 30_000 : 20 * 60_000 });
  try {
    // Validate the archive and exact WAV before making a directory or loading CUDA.
    await execute("reference-text", path.join(root, "unused-check-output"), true);
    if (values["check-only"]) {
      console.log("Diagnostic et référence concordants. Aucun modèle chargé, aucun calcul lancé.");
      return;
    }
    const parent = path.join(root, "diagnostics", "comparisons");
    await fs.promises.mkdir(parent, { recursive: true, mode: 0o700 });
    if (values["prepare-only"]) {
      const bundle = await fs.promises.mkdtemp(path.join(parent, `input-${label}-`));
      const names = ["quality-failure.json", "error.json", "generation-attempt.json", "input-profile.json",
        ...(english ? ["quality-attempt-1.wav", "quality-attempt-2.wav", "quality-attempt-3.wav"] : [])];
      for (const name of names.filter(name => fs.existsSync(path.join(failure, name)))) {
        await fs.promises.copyFile(path.join(failure, name), path.join(bundle, name));
        await fs.promises.chmod(path.join(bundle, name), 0o600);
      }
      await fs.promises.copyFile(reference, path.join(bundle, "reference.wav"));
      await fs.promises.chmod(path.join(bundle, "reference.wav"), 0o600);
      console.log(`Entrées privées prêtes à copier sur le clone : ${bundle}\nUtiliser ce dossier pour --failure et son fichier reference.wav pour --reference. Aucun calcul GPU lancé.`);
      return;
    }
    const output = await fs.promises.mkdtemp(path.join(parent, `${label}-`));
    console.log(`Diagnostics privés : ${output}`);
    const summary = { schemaVersion: 1, synthetic: true, diagnosticOnly: true, language, variants: [], requiresListening: true };
    let failed = false;
    for (const variant of variants) {
      if (interrupted) break;
      console.log(`Comparaison ${variant} : 3 tentatives sur une seule réplique.`);
      try {
        await execute(variant, path.join(output, variant));
        const reportPath = path.join(output, variant, "comparison.json");
        const report = fs.existsSync(reportPath) ? JSON.parse(await fs.promises.readFile(reportPath, "utf8")) : null;
        const rejected = report?.attempts?.filter(attempt => attempt.state === "rejected-before-decoding").length || 0;
        summary.variants.push({ variant, state: rejected ? "completed-with-rejections" : "completed",
          rejectedBeforeDecoding: rejected, report: `${variant}/comparison.json` });
        if (rejected) console.log(`${variant} : ${rejected}/3 tentative(s) refusée(s) avant décodage ; aucun WAV pour ces tentatives, détails dans comparison.json.`);
      } catch (error) {
        failed = true;
        summary.variants.push({ variant, state: "failed", message: error.message, code: error.code });
        console.error(error.message);
      }
      await fs.promises.writeFile(path.join(output, "summary.json"), JSON.stringify(summary, null, 2), { mode: 0o600 });
    }
    if (failed || interrupted) process.exitCode = 1;
    console.log(`Comparaison ${failed || interrupted ? "incomplète" : "terminée"}. Écouter les WAV AI-diagnostic et consulter comparison.json ; rien n'a été publié dans SAMI.`);
  } finally {
    process.removeListener("SIGINT", stop);
    process.removeListener("SIGTERM", stop);
  }
}

if (process.argv[1] && path.resolve(process.argv[1]) === fileURLToPath(import.meta.url)) {
  main().catch(error => { console.error(error.message); process.exitCode = 1; });
}
